src/mlpack/core/data/one_hot_encoding_impl.hpp

```diff
@@ -56,14 +56,51 @@ void OneHotEncoding(const RowType& labelsIn,
     }
   }
   // Resize output matrix to necessary size, and fill it with zeros.
-  output.zeros(labelsIn.n_elem, curLabel);
+  output.zeros(curLabel, labelsIn.n_elem);
   // Fill ones in at the required places.
   for (size_t i = 0; i < labelsIn.n_elem; ++i)
   {
-    output(i, labels[i]) = 1;
+    output(labels[i], i) = 1;
   }
   labelMap.clear();
 }
+
+/**
+ * Overloaded function for the above function, which takes a matrix as input
+ * and also a vector of indices to encode and outputs a matrix.
+ *
+ * @param input Input dataset to be encoded.
+ * @param indices Index of rows to be encoded.
+ * @param output Encoded matrix.
+ */
+template<typename eT>
+void OneHotEncoding(arma::Mat<eT>& input,
+                    const arma::ucolvec indices,
+                    arma::Mat<eT>& output)
+{
+  output = input;
+  size_t rowCount = 0;
+  for (size_t i = 0; i < indices.n_elem; i++)
+  {
+    output.shed_row(indices.at(i) - rowCount);
+    rowCount++;
+  }
+
+  std::vector<arma::Mat<eT>> oheOutput(indices.n_elem);
+  for (size_t i = 0; i < indices.n_elem; i++)
+  {
+    // Call OneHotEncoding() for each of the indices.
+    OneHotEncoding(input.row(indices.at(i)), oheOutput[i]);
+  }
+  size_t row = 0;
+  for (size_t i = 0; i < indices.n_elem; i++)
+  {
+    // Calculate the index at which the rows should be inserted.
+    output.insert_rows(row + indices.at(i), oheOutput[i]);
+    row += oheOutput[i].n_rows - 1;
+  }
+}
```

Comment:
This algorithm is actually kind of inefficient.  I would suggest that it would be better to do this in two passes: first take a pass over `input` and determine how many dimensions should be in the `output` matrix.  Then, set `output` to the right size, fill it with zeros, and then pass over `input` once more to fill the values in `output`.  This will be a lot better than calling `shed_row()` (which will internally copy the whole matrix) once for each column in `indices`.  `insert_rows()` has the same problem where it will copy the entire matrix.  So, if you can do this entire function (a) without storing a `std::vector<arma::Mat<eT>>`, (b) without ever calling `shed_row()`, and (c) without ever calling `insert_rows()`, I think it will be a much better implementation. :+1:

Reply:
Done

Reply:
Awesome, thanks.  I have some more comments but I'll resolve this one. :)